CONTRIBUTING.md

```diff
@@ -570,6 +570,26 @@ It's a "to do" item and should be replaced if possible. `Any` is used when
 it's not possible to accurately type an item using the current type system.
 It should be used sparingly.
 
+### "The `Any` trick"
+
+Consider this: (mostly copy/paste from xml module docs)
+
+```python
+import xml.etree.ElementTree as ET
+tree = ET.parse('country_data.xml')
+root = tree.getroot()
+print(root.tagg)  # this line is a typo
+```
```

Comment:
I'm actually having a hard time coming up with a simple small example with `re.match` where it makes sense to return `re.Match[T] | Any` instead of just `re.Match[str]` if the string is already validated.

Like:
```py
import re

PATTERN = re.compile("test")

def match_my_pattern(value: str) -> re.Match[str]:
    assert value.startswith("test")
    return re.match(PATTERN, value)
```

I can't find any example of `-> Match[T] | Any` in typeshed either.

Reply:
Ah right, we use the trick for `Match.group`: https://github.com/python/typeshed/blob/acfdebfeb83c110d8ffefcfa7cc68bf3ab94817f/stdlib/re.pyi#L77

Not as prominent as `re.match` itself, but might still be a better example than xml.

Reply:
I think the explanation is now a lot clearer as-is. The code example could probably be improved.

Reply:
Maybe we should go with something really practical for the example. For example, suppose you have a legacy system that for historical reasons has two kinds of user IDs. Old IDs look like `"legacy_userid_123"` and new IDs look like `"456_akuli"`. The function below extracts the number from a new ID and returns `None` if you give it a legacy ID.

```python
def parse_name_from_new_id(user_id: str) -> int | None:
    match = re.fullmatch(r"\d+_(.*)", user_id)
    if match is None:
        return None
    return match.group(1).capitalise()
```

Reply:
Here's another example, mostly taken from [editorconfig parsing code in one of my projects](https://github.com/Akuli/porcupine/blob/main/porcupine/plugins/editorconfig.py#L147-L164). It finds the 1 and 10 from a string that looks like like `"foo{1..10}.txt"`.

```python
match = re.fullmatch(r"\{(-?[0-9]+)\.\.(-?[0-9]+)\}", glob)
if match is not None:
    min_value = int(match.group(1))
    max_value = int(match.group(2))
```

Reply:
I'm less fond of your 2nd example, because it doesn't show that the trick still catches invalid attributes (due to a type for example).

I'll update your first example with the more obvious `uper` typo.